src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -0,0 +1,518 @@
+        }
+
+        thang = thangs.get(item)
+        if thang is None:
+            raise RuntimeError("Don't know how to wait for {item} {operation}")
+        elapsed = 0.0
+        while elapsed < self.timeout:
+            errored = False
+            try:
+                await asyncio.wait_for(
+                    thang,
+                    self.timeout,
+                )
+                return
+            except ApiException as e:
+                if e.status != 404:
+                    if operation == ObjectOperation.DELETION:
+                        return
+                    raise WaitingForObjectError(str(e))
+                errored = True
+            if not errored and operation == ObjectOperation.CREATION:
+                return
+            # If we're here, it's a delete, and the read didn't get an
+            # error, or it's a creation and we did get an error, so we
+            # need to wait and then retry
+            await asyncio.sleep(interval)
+            elapsed += interval
+        raise asyncio.TimeoutError(
+            f"Timeout while waiting for {item} {operation}"
+        )
+
+    async def copy_pull_secret(self, source_ns: str) -> Dict[str, str]:
+        secret = await self.read_secret(
+            name="pull-secret", namespace=source_ns
+        )
+        return secret.data
+
+    async def merge_controller_secrets(
+        self, secret_list: List[LabSecret], token: str, source_ns: str
+    ) -> Dict[str, str]:
+        """Merge the user token with whatever secrets we're injecting
+        from the lab controller environment."""
+        secret_names: List[str] = list()
+        secret_keys: List[str] = list()
+        for sec in secret_list:
+            secret_names.append(sec.secret_name)
+            if sec.secret_key in secret_keys:
+                raise RuntimeError("Duplicate secret key {sec.secret_key}")
+            secret_keys.append(sec.secret_key)
+        # In theory, we should parallelize the secret reads.  But in practice
+        # it makes life a lot more complex, and we probably just have one,
+        # the controller secret.  Pull-secret will be handled separately.
+        base64_data: Dict[str, str] = dict()
+        for name in secret_names:
+            secret: Secret = await self.read_secret(
+                name=name, namespace=source_ns
+            )
+            # Retrieve matching keys
+            for key in secret.data:
+                if key in secret_keys:
+                    base64_data[key] = secret.data[key]
+        # There's no point in decoding it; all we're gonna do is pass it
+        # down to create a secret as base64 anyway.
+        if "token" in base64_data:
+            raise RuntimeError("'token' must come from the user token")
+        if not token:
+            raise RuntimeError("User token cannot be empty")
+        base64_data["token"] = str(base64.b64encode(token.encode("utf-8")))
```

Comment:
```suggestion
        base64_data["token"] = str(base64.b64encode(token.encode()))
```
UTF-8 is the default and only sensible encoding for most programs, so you can leave it out.